rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -2822,6 +2867,12 @@ static rmw_ret_t rmw_take_int(
   dds_sample_info_t info;
   while (dds_take(sub->enth, &ros_message, &info, 1, 1) == 1) {
     if (info.valid_data) {
+      if (sub->is_loaning_available) {
+        // if the sample has been transferred using iceoryx, the sample from iceoryx chunk is
+        // copied into ros_message in `_to_sample()` and the chunk is released here
+        dds_data_allocator_init(sub->enth, &sub->data_allocator);
+        dds_data_allocator_fini(&sub->data_allocator);
```

Comment:
You are correct, I did miss freeing the chunk. This commit should fix this https://github.com/ros2/rmw_cyclonedds/pull/309/commits/5e64cac8f34434560cc929cad50119c98f7e0e2d

Reply:
I'm not sure how serializing and deserializing the data frees the chunk, but it sounds ok to me if that works (though it doesn't seem to be an ideal approach performance wise)

Reply:
The chunk should be release in serdata_free. Now we are not copying the data from chunk to the sample, but do the serialization on demand, and the data path should be the same as previous. \
We can argue about the performance, If the data is published with SHM, but a regular `rmw_take()` is invoked, then probably we have the following 2 options:
1. We can copy the data from the chunk to the ros message (since the chunk already has deserialized sample) or
2. We can serialize the chunk and the regular data path continues.
I choose option-2 since it causes minimum disruption to the existing data path. Option-1 is slightly more performant, but that means it should handle different scenarios as following:
1. Use `take_cdr()` (from cyclone) instead of regular `take()`, to avoid serialization and deserialization with SHM
2. If the sample is published using SHM, but a regular `rmw_take()` is invoked
    1. Copy the shared memory chunk to the ros message
3. If the sample is published regularly, but a regular `rmw_take()` is invoked with SHM enabled
   1. De-serialize the data to the ros message
  
The above approach, should the most performant data path. However, it disrupts the current data path since we will have 2 conditional branches in `rmw_take()` one with `take_cdr()` (from cyclone) and the other with `take()`, which are executed based on if loaning is available or not.

If you strongly feel that we should change this, then I can change it in this MR or in the following issue.

Reply:
Thanks for providing extra context @sumanth-nirmal .

> If you strongly feel that we should change this, then I can change it in this MR or in a following issue.

I don't feel strongly.
I was confused of how the changes in `serdata.cpp` in https://github.com/ros2/rmw_cyclonedds/commit/5e64cac8f34434560cc929cad50119c98f7e0e2d were related to the fact that [these lines](https://github.com/ros2/rmw_cyclonedds/blob/712fd5d413c3059b8c96b9160d8f8ae910c7886c/rmw_cyclonedds_cpp/src/rmw_node.cpp#L2870-L2875) weren't needed.
If this is already freeing the chunk correctly, the the code looks good to me.